File: Data Structures & Algorithms/climbing-stairs/submission-0.go
func climbStairs(n int) int {
    cache := make([]int, n+1)

    for i := range cache {
        cache[i] = -1
    }



    var step func(int) int
    step = func(at int) int {
        if at == n {
            return 1
        }
        if at > n {
            return 0
        }

        if cache[at] != -1 {
            return cache[at]
        }

        stepAt := step(at+1) + step(at+2)
        cache[at] = stepAt
        return stepAt
    }
    res := step(0)
    return res
}

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-0.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
       # subproblem is: at each step, we calculate the minimum cost to reach this stair case. 
       # suppose cache is like this[1,4,5,7,6], res = min of last two values

        length = len(cost)
        cache = [0] * length

        cache[0] = cost[0]
        cache[1] = cost[1]

        for i in range(2, length):
            cache[i] = cost[i] + min(cache[i-1], cache[i-2])

        print(cache)
        return min(cache[length-1], cache[length-2])
